Add `DatePicker` tests using React Testing Library

noisysocks opened the pull request:
## What?

Rewrites the `DatePicker` tests to use React Testing Library and to test the component's basic functionality instead of simply that `react-dates` is called correctly.

## Why?

- Gives us some confidence that the basic functionality of `DatePicker` works. (I'm about to refactor it.)
- Makes the tests align more closely with `DatePicker`'s sister component, `TimePicker`.

## How?

Wrote new tests using React Testing Library. I focused on ensuring coverage for all of the public props.

I didn't touch `TimePicker`'s tests as they already look pretty good.

I also added a new story that demonstrates the `isInvalidDate` prop which was missing.

## Testing Instructions

`npm run test`

noisysocks committed:
Add DatePicker tests using React Testing Library

mirka commented on packages/components/src/date-time/stories/index.js:
```diff
@@ -70,3 +70,17 @@ export const WithDaysHighlighted = () => {
 	);
 };
+
+export const WithInvalidDates = () => {
```
In tandem with #35665 and #35744, we're in the process of making our Storybook work more as standalone documentation ([example end result](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-unitcontrol--default)). With that end goal in mind, we want to make sure that any stories we add are useful and understandable as part of public documentation.

It's a bit unclear to me what this story is demonstrating, so I think it warrants a JSDoc comment to describe what it's trying to show (the JSDoc comment will show up in the Storybook Docs view). And if it's actually just a test fixture, we can remove the story.

noisysocks replied:
Ooh that's nice! I'll add a JSDoc comment. This story just demonstrates how to use the `isInvalidDate` prop. Definitely this component's stories could do with a larger update to be more coherent down the line.

mirka commented on packages/components/src/date-time/test/date.js:
```diff
@@ -1,97 +1,111 @@
  */
-import { shallow } from 'enzyme';
 import moment from 'moment';
+import { render, screen, fireEvent } from '@testing-library/react';
```
FYI we now have [`user-event`](https://testing-library.com/docs/user-event/intro) in this repo, so it's probably better for us to use `userEvent` instead of `fireEvent` by default.

mirka replied:
FYI we now have [`user-event`](https://testing-library.com/docs/user-event/intro) in this repo, so it's probably better for us to use `userEvent` instead of `fireEvent` by default.

mirka reviewed (commented):
Wonderful, thanks for improving the tests!

noisysocks committed:
Add comments to WithInvalidDates story

noisysocks committed:
Use @testing-library/user-event

noisysocks commented:
Thanks @mirka. I added a story comment and switched to `userEvent`. I had to configure `delay: null` when initialising `userEvent` because otherwise the tests would timeout. I think it's the same issue as what @ciampo ran into in https://github.com/testing-library/user-event/issues/565#issuecomment-1064579531. Let me know if you have a better idea of how to fix that (or feel free to push to the branch).

kevin940726 commented on packages/components/src/date-time/test/date.js:
```diff
@@ -1,97 +1,118 @@
+
+		expect(
+			screen.getByLabelText( 'Monday, May 2, 2022' ).classList
+		).toContain( 'CalendarDay__selected' );
```
I wonder if we could do better than asserting class names which are implementation details. Does `react-dates` add any accessible properties to the selected date?

noisysocks replied:
Yeah, I tried to follow [the advice](https://testing-library.com/docs/queries/about#priority) on using accessible queries as much as I could but unfortunately had to use a class name for this as it's all that `react-dates` gives us.

<img width="696" alt="Screen Shot 2022-05-04 at 09 58 30" src="https://user-images.githubusercontent.com/612155/166604713-2d1b992a-b967-4825-9abc-aadbfad865b5.png">

If you click on a date then that date gets `Selected. ` prepended to the label but for some reason `react-dates` doesn't do this for the initial selection which is what we want to test here.

Ideally `react-dates` would use `aria-pressed` or similar. Maybe it does in newer versions (**edit:** nope, it does not). At any rate I think it's out of scope for this PR.

Let me know if you spot a better approach.

noisysocks replied:
(Personally I think `react-dates` needs to be changed to use `aria-selected` or something like that but that this is out of scope for this PR 😀)

kevin940726 replied:
Yeah agree. It should be fixed in `react-dates`' end. Thanks for investigating!

kevin940726 replied:
Another option would be to use `jest-dom`'s [`toHaveStyle`](https://github.com/testing-library/jest-dom#tohavestyle), but that's also testing implementation detail. I think we can at least use [`toHaveClass`](https://github.com/testing-library/jest-dom#tohaveclass) though, WDYT? (Assuming `jest-dom` is already enabled?)

noisysocks replied:
Didn't know about that one! Neat 👍

kevin940726 commented on packages/components/src/date-time/test/date.js:
```diff
@@ -1,97 +1,118 @@
+		render( <DatePicker currentDate="2022-05-02T11:00:00" /> );
+
+		expect(
+			screen.getByLabelText( 'Monday, May 2, 2022' ).classList
```
Nit: `getByRole` could be slightly better than `getByLabelText` because it makes it more readable for which aria role the element has.

noisysocks replied:
Oops, missed that these had a role! 👍

aaronrobertshaw reviewed (commented):
Nice work @noisysocks 👍 

✅ Storybook example looks good until we can migrate [Knobs to Controls
](https://github.com/WordPress/gutenberg/issues/35665)✅ Unit tests pass and make sense to me

❓ I also had the same [question](https://github.com/WordPress/gutenberg/pull/40754#discussion_r863514668) regarding whether we could assert date selection based off of accessibility properties.

I'm not very familiar with `react-dates` and didn't find anything to solve that. Perhaps you'll have more ideas.

ciampo commented on packages/components/src/date-time/test/date.js:
```diff
@@ -1,97 +1,118 @@
-			).toBe( true );
-		} );
+	it( 'should call onMonthPreviewed and onChange when a day in a different month is selected', async () => {
+		const user = userEvent.setup( { delay: null } );
```
Would you mind making a few changes to these tests in order to align to how we use `@testing-library/user-event` in other components?

You'd need to rebase on top of `trunk` to include [this version bump to `user-event`](https://github.com/WordPress/gutenberg/pull/40761) (don't forget to reinstall `node_modules`), and then switch to use jest fake timers + the `advanceTimers` option from `userEvent`

<details>
<summary>Example</summary>

```diff
diff --git a/packages/components/src/date-time/test/date.js b/packages/components/src/date-time/test/date.js
index f204dff80f..4cf66697f4 100644
--- a/packages/components/src/date-time/test/date.js
+++ b/packages/components/src/date-time/test/date.js
@@ -11,7 +11,20 @@ import 'react-dates/initialize';
  */
 import DatePicker from '../date';
 
+const user = userEvent.setup( {
+	advanceTimers: jest.advanceTimersByTime,
+} );
+
 describe( 'DatePicker', () => {
+	beforeEach( () => {
+		jest.useFakeTimers();
+	} );
+
+	afterEach( () => {
+		jest.runOnlyPendingTimers();
+		jest.useRealTimers();
+	} );
+
 	it( 'should highlight the current date', () => {
 		render( <DatePicker currentDate="2022-05-02T11:00:00" /> );
 
@@ -34,8 +47,6 @@ describe( 'DatePicker', () => {
 	} );
 
 	it( 'should call onChange when a day is selected', async () => {
-		const user = userEvent.setup( { delay: null } );
-
 		const onChange = jest.fn();
 
 		render(
@@ -51,8 +62,6 @@ describe( 'DatePicker', () => {
 	} );
 
 	it( 'should call onMonthPreviewed and onChange when a day in a different month is selected', async () => {
-		const user = userEvent.setup( { delay: null } );
-
 		const onMonthPreviewed = jest.fn();
 		const onChange = jest.fn();
 
@@ -99,8 +108,6 @@ describe( 'DatePicker', () => {
 	} );
 
 	it( 'should not allow invalid date to be selected', async () => {
-		const user = userEvent.setup( { delay: null } );
-
 		const onChange = jest.fn();
 
 		render(
```
</details>

I just merged a PR with similar changes today (#40790).

Also happy if you want to carry these changes in a follow-up !

noisysocks replied:
Opened a follow-up for this: https://github.com/WordPress/gutenberg/pull/40800

Unfortunately the change causes a weird error in `react-dates`. Let's discuss in the other PR.

ciampo reviewed (commented):
Agree with the points made by @kevin940726 and @aaronrobertshaw .

Would it be also possible to add an entry to the CHANGELOG (this one could be under `Internal`)

noisysocks committed:
Use getByRole

noisysocks committed:
Add CHANGELOG entry

noisysocks commented:
I've added a `CHANGELOG` entry and switched to using `getByRole`. I don't see a nicer way of asserting that the button is selected, unfortunately, but let me know if you do. I'll handle the changes to `userEvent.setup()` in a follow-up PR (https://github.com/WordPress/gutenberg/pull/40800) as doing this results in an error that will need to investigate.

kevin940726 reviewed (approved):
LGTM!

Just a small nitpicking left [here](https://github.com/WordPress/gutenberg/pull/40754/files#r864453738), but overall I think this is good enough 💯 .

noisysocks committed:
Use toHaveClass
